### PR TITLE
SqlClient: Reduce magic strings

### PR DESCRIPTION
**Pull request by justinvp**
Use `nameof()` and `[CallerMemberName]` to reduce magic strings in SqlClient. Also delete unused string constants.

Part of #7290.

cc: @saurabh500

### CODE REVIEW

File: src/System.Data.SqlClient/src/Microsoft/SqlServer/Server/SqlDataRecord.cs [hunk 1]
```diff
@@ -649,7 +649,7 @@ public SqlDataRecord(params SqlMetaData[] metaData)
             // Initial consistency check
             if (null == metaData)
             {
-                throw ADP.ArgumentNull("metadata");
+                throw ADP.ArgumentNull(nameof(metaData));
```
Comment (justinvp) [inline src/System.Data.SqlClient/src/Microsoft/SqlServer/Server/SqlDataRecord.cs]: Deliberate change to match the casing of the parameter.

File: src/System.Data.SqlClient/src/Microsoft/SqlServer/Server/SqlDataRecord.cs [hunk 2]
```diff
@@ -658,7 +658,7 @@ public SqlDataRecord(params SqlMetaData[] metaData)
             {
                 if (null == metaData[i])
                 {
-                    throw ADP.ArgumentNull("metadata[" + i + "]");
+                    throw ADP.ArgumentNull($"{nameof(metaData)}[{i}]");
```
Comment (justinvp) [inline src/System.Data.SqlClient/src/Microsoft/SqlServer/Server/SqlDataRecord.cs]: Deliberate change to match the casing of the parameter.

File: src/System.Data.SqlClient/src/System/Data/Common/AdapterUtil.cs [hunk 3]
```diff
@@ -827,69 +828,14 @@ static internal Exception TransactionZombied(DbTransaction obj)
...
-        internal const string RollbackTransaction = "RollbackTransaction";
-        internal const string SaveTransaction = "SaveTransaction";
-        internal const string SetProperties = "SetProperties";
-        internal const string SourceColumn = "SourceColumn";
-        internal const string SourceVersion = "SourceVersion";
-        internal const string SourceTable = "SourceTable";
-        internal const string UnquoteIdentifier = "UnquoteIdentifier";
-        internal const string Update = "Update";
-        internal const string UpdateCommand = "UpdateCommand";
-        internal const string UpdateRows = "UpdateRows";
```
Comment (justinvp) [inline src/System.Data.SqlClient/src/System/Data/Common/AdapterUtil.cs]: All the removed constants above were either dead code before this PR or became dead code as part of this PR by switching uses to `nameof()` or `[CallerMemberName]`.

File: src/System.Data.SqlClient/src/System/Data/Common/AdapterUtil.cs [hunk 4]
```diff
@@ -403,7 +404,7 @@ static internal InvalidOperationException NoConnectionString()
             return InvalidOperation(Res.GetString(Res.ADP_NoConnectionString));
         }
 
-        static internal Exception MethodNotImplemented(string methodName)
+        static internal Exception MethodNotImplemented([CallerMemberName] string methodName = "")
```
Comment (saurabh500) [inline src/System.Data.SqlClient/src/System/Data/Common/AdapterUtil.cs]: @justinvp Why did you change these?

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlStream.cs [hunk 5]
```diff
@@ -469,27 +469,27 @@ override public long Seek(long offset, SeekOrigin origin)
 
             if (null == _cachedBytes)
             {
-                throw ADP.StreamClosed(ADP.Read);
+                throw ADP.StreamClosed();
```
Comment (justinvp) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlStream.cs]: I deliberately changed this from the previously incorrect "Read" to the correct "Seek" (via `[CallerMemberName]`).

File: src/System.Data.SqlClient/src/System/Data/Common/AdapterUtil.cs (hunk 4 above)
Comment (justinvp) [inline src/System.Data.SqlClient/src/System/Data/Common/AdapterUtil.cs]: So callers can call it like `MethodNotImplemented();` instead of `MethodNotImplemented("Caller");` and the compiler will automatically fill in `"Caller"`.

Comment (saurabh500) [inline src/System.Data.SqlClient/src/System/Data/Common/AdapterUtil.cs]: Commented before looking at the rest of the Diff.

**Comment by saurabh500** [comment]
@justinvp Thanks for the changes.  I am going through them. 

cc @nadiatk @corivera @kkurni

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs [hunk 6]
```diff
@@ -889,13 +889,13 @@ private void BeginExecuteNonQueryInternalReadStage(TaskCompletionSource<object>
             }
         }
 
-        private void VerifyEndExecuteState(Task completionTask, String endMethod)
+        private void VerifyEndExecuteState(Task asyncResult, String endMethod)
```
Comment (stephentoub) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs]: Seems a bit backwards to change the name of the parameter from the more applicable "completionTask" to the less applicable "asyncResult", simply to be able to use ```nameof(asyncResult)``` instead of ```"asyncResult"```.

**Comment by stephentoub** [comment]
One nit, otherwise LGTM.

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs (hunk 6 above)
Comment (justinvp) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs]: `asyncResult` seems like just as reasonable a name as `completionTask`, in my opinion. `VerifyEndExecuteState` is only used in two places to verify the asyncResult, which happens to be a Task.

Comment (stephentoub) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs]: It's strongly typed as a Task.  It's not like it's an IAsyncResult that happens to be a Task and is being cast to one internally.

Comment (saurabh500) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs]: I agree with @stephentoub on this. Otherwise the changes look good to me.

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs [hunk 7]
```diff
@@ -748,7 +748,7 @@ override public object ExecuteScalar()
             {
                 statistics = SqlStatistics.StartTimer(Statistics);
                 SqlDataReader ds;
-                ds = RunExecuteReader(0, RunBehavior.ReturnImmediately, true, ADP.ExecuteScalar);
+                ds = RunExecuteReader(0, RunBehavior.ReturnImmediately, true, nameof(ExecuteScalar));
```
Comment (saurabh500) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs]: Wouldn't it make more sense to change RunExecuteReader to take [CallerMemberName] string method=""
You can get rid of most of the nameof(...) being passed down to RunExecuteReader as well.

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs (hunk 6 above)
Comment (justinvp) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs]: Since the public `Begin`/`End` methods don't exist in .NET Core, the `Task` passed in here should never be null, so I've replaced the `ArgumentNullException` with a `Debug.Assert` to avoid the parameter rename altogether.

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs (hunk 7 above)
Comment (justinvp) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs]: Good point. Updated.

**Comment by justinvp** [comment]
Thanks for the feedback, @saurabh500 and @stephentoub. I added a commit that addresses your comments. (I temporarily added back a string constant to get rid of the public `EndExecuteNonQuery` method that is not in the public contract, to be reconciled in the subsequent commit).

I also added another commit that replaces the remaining string constants with uses of CallerMemberName and nameof, by removing the Async suffix from the private BeginXxxAsync and
EndXxxAsync methods. Please take a look.

**Comment by justinvp** [comment]
Test Innerloop Windows_NT Debug Build and Test please

(#4467 and #7517)

**Comment by justinvp** [comment]
Test Innerloop CentOS7.1 Release Build and Test please

**Comment by stephentoub** [comment]
Thanks, Justin.
